File: server/app/models/discount.py
from app.utils.format_time import to_utc_iso

def discount_model(discount: dict) -> dict:
    return {
        "_id": str(discount["_id"]),
        "name": discount["name"],
        "code": discount.get("code"),
        "description": discount.get("description"),
        "image": discount.get("image"),
        "isFeature": discount.get("isFeature", False),
        "type": discount["type"],
        "value": discount["value"],
        "minOrderValue": discount.get("minOrderValue", 0),
        "maxDiscountValue": discount.get("maxDiscountValue"),
        # Always return simplified fields
        "applyTo": "order",
        "applyToAllCategories": True,
        "applicableProducts": [],
        "applicableCategories": [],
        "maxUsageCount": discount.get("maxUsageCount", 0),
        "usageCount": discount.get("usageCount", 0),
        "maxUsagePerUser": discount.get("maxUsagePerUser", 0),
        "userCondition": discount.get("userCondition", "all"),
        "startDate": to_utc_iso(discount.get("startDate")),
        "endDate": to_utc_iso(discount.get("endDate")),
        "isActive": discount.get("isActive", True),
        "stackable": discount.get("stackable", False),
    }

def build_applied_discount(discount: dict, value: float, mode: str):
    return {
        "_id": str(discount["_id"]),
        "name": discount.get("name"),
        "code": discount.get("code"),
        "type": discount.get("type"),
        "value": discount.get("value"),
        "discountValue": round(value, 2),
        "mode": mode,          # auto | manual
        "source": mode         # FE dùng source cho UI
    }

def discounts_model(discounts_cursor) -> list[dict]:
    return [discount_model(discount) for discount in discounts_cursor]


def discount_available(discount):
    return {
        "_id": discount["_id"],
        "code": discount["code"],
        "name": discount["name"],
        "description": discount["description"],
        "type": discount["type"],
        "value": discount["value"],
        "maxDiscountValue": discount["maxDiscountValue"],
        "minOrderValue": discount["minOrderValue"],
        "usageCount": discount["usageCount"],
        "maxUsageCount": discount["maxUsageCount"],
        "maxUsagePerUser": discount["maxUsagePerUser"],
        "stackable": discount["stackable"],
        "eligible": discount.get("eligible", True),
        "reason": discount.get("reason", ""),
        "requireLoginToUse": discount.get("requireLoginToUse", False),
        "estimatedDiscount": discount.get("estimatedDiscount", 0),
    }

File: client-user/types/discount.ts
import { CategoryForSelect } from "./category";

// Discount.types.ts
export interface DiscountAdmin {
    _id: string;                    // ObjectId as string
    code: string;                  // mã khuyến mãi duy nhất
    name: string;                   // tên khuyến mãi
    description: string;           // mô tả chi tiết
    type: DiscountType;            // loại khuyến mãi
    image: string;                 // URL ảnh khuyến mãi

    // Điều kiện áp dụng
    value: number;                  // giá trị giảm (%, số tiền cố định...)
    minOrderValue: number;         // đơn tối thiểu để áp dụng
    maxDiscountValue: number;     // giới hạn giảm tối đa (nếu là %)

    applyTo: DiscountApplyTo;
    applyToAllCategories: boolean;

    // Thời gian
    startDate: string;      // ISO string hoặc null (vô thời hạn bắt đầu)
    endDate: string;        // ISO string hoặc null (vô thời hạn kết thúc)

    // Giới hạn sử dụng
    maxUsageCount: number;         // tổng lượt dùng tối đa (0 = không giới hạn)
    usageCount: number;             // đã dùng bao nhiêu lần
    maxUsagePerUser: number;          // lượt dùng tối đa mỗi user (0 = không giới hạn)

    // Trạng thái & flags
    isActive: boolean;
    isFeature: boolean;
    status: DiscountStatus;        // trạng thái hiện tại (từ scheduler hoặc realtime)

    // Các trường bổ sung tùy dự án
    applicableProducts: { _id: string, name: string }[];  // danh sách product ID áp dụng (nếu không phải all)
    applicableCategories: CategoryForSelect[];
    excludedProducts: string[];

    priority: DiscountPriority;
    userCondition: UserCondition;        // điều kiện người dùng (mới, VIP, tất cả)
    stackable: boolean;            // có thể dùng chung với khuyến mãi khác không

    createdAt: string;
    updatedAt: string;
}

export interface Discount {
    _id: string;
    // Hiển thị
    name: string;
    description: string;
    image: string;
    isFeature: boolean;

    // Giảm giá
    type: DiscountType;          // percentage | fixed | shipping
    value: number;
    maxDiscountValue?: number;
    minOrderValue?: number;

    maxUsageCount: number;         // tổng lượt dùng tối đa (0 = không giới hạn)
    usageCount: number;             // đã dùng bao nhiêu lần
    maxUsagePerUser: number;

    // Phạm vi áp dụng (để FE check nhanh)
    applyTo: DiscountApplyTo;
    applicableProducts?: { _id: string; name: string }[];
    applicableCategories?: CategoryForSelect[];

    // Điều kiện cơ bản
    userCondition?: UserCondition;

    // Thời gian (để FE hiển thị badge)
    startDate: string;
    endDate: string;

    // Cờ trạng thái đơn giản
    isActive: boolean;
    stackable: boolean;
}

export interface AppliedDiscount {
    _id: string;
    code: string;
    name: string;
    description: string;
    amount: number;
    type: DiscountType;
    value: number;
}

export type DiscountAvailable = {
  _id: string
  code: string
  name: string
  description: string

  type: DiscountType
  value: number

  maxDiscountValue: number 
  minOrderValue: number 

  usageCount: number
  maxUsageCount: number
  maxUsagePerUser: number

  stackable: boolean

  eligible: boolean
  reason: string
  requireLoginToUse: boolean

  estimatedDiscount: number
}

// type cho type và status
export type DiscountType = "percentage" | "fixed" | "buy_x_get_y" | "shipping"

export type DiscountStatus = "active" | "scheduled" | "expired" | "exhausted" | "inactive"

export type UserCondition = "all" | "new" | "vip"

export type DiscountPriority = "low" | "medium" | "high"

export type DiscountApplyTo = "order" | "product"